### PR TITLE
UCT/UD: Add support for zero-copy control messages

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -1005,28 +1008,32 @@ static uct_ud_send_skb_t *uct_ud_ep_resend(uct_ud_ep_t *ep)
 static void uct_ud_ep_do_pending_ctl(uct_ud_ep_t *ep, uct_ud_iface_t *iface)
 {
     uct_ud_send_skb_t *skb;
-    int flag = 0;
+    int is_creq_crep = 0;
+    int ctl_flags = 0;
```
Comment (dmitrygx) [inline src/uct/ib/ud/base/ud_ep.c]:
```
int is_creq_crep = 0;
int ctl_flags    = 0;
```

File: src/uct/ib/ud/verbs/ud_verbs.c [hunk 2]
```diff
@@ -65,73 +65,97 @@ static UCS_CLASS_DEFINE_NEW_FUNC(uct_ud_verbs_ep_t, uct_ep_t,
...
+                       struct ibv_send_wr *wr, unsigned send_flags,
+                       unsigned max_log_sge)
 {
-    if (iface->super.tx.unsignaled >= UCT_UD_TX_MODERATION) {
-        wr->send_flags       = (flags|IBV_SEND_SIGNALED);
+    struct ibv_send_wr *bad_wr;
+    int UCS_V_UNUSED ret;
+
+    if ((send_flags & IBV_SEND_SIGNALED) ||
+        (iface->super.tx.unsignaled >= UCT_UD_TX_MODERATION - 1)) {
```
Comment (dmitrygx) [inline src/uct/ib/ud/verbs/ud_verbs.c]: `UCT_UD_TX_MODERATION - 1`

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 01-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/bd9b664acbd44331f2dbb5eccff55318908db468/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/8949035736b40ef025b2cc9166e0c449ef990a7e/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/e32af52be902280cfc35ac1a96ded4d9233a4be2/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/7e799e5251f4ca54fd08d43f32f4be7e0cfeda01/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/6096096c4c8fabdef41caaa08e4931606ea856d6/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/1fba30d7191912f2ae442d79210389236f69b0a6/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/504802d5a06238c23b17867e450789ef0befd3a2/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/e02c3cd2da496ec4e3d102db0125907427415988/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/7f03b393618ba4f77dbfd3148242167f2d5bb02c/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/2bde422a5b3753f6d4dc118d93f96a38ece69a83/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/87b4271fc69e9230a3f67dc3b62015608db3cbd8/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/19fe518d3c39ec1af2b462a8a4ef26ae8a7f2c29/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/ed04663412cfced938ab5ab543ec2466e892c402/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/9feb18871dead59cc36768e37ca030c371974212/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/356af189f056f80ddcfa0c95d7a24fedd5c5b607/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/784b6132b529c313e0bccef671ae11d56e1afc46/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/095932be353b5dca28025a1e21a25ba26ff4519e/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/af88a4e21e018002b4c0b5c07107b95594de5bfd/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/072fa5a0d4cc9f33a74c7d6e150b309f970c5cc4/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/3fcb6745a0eea9b448bb0c934ef8438a74f7d22d/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/29409e1d376686d07903849c45008a6d161e4013/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/67154a46f096ff1ae1b50b19bbab413dbe8aeff4/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/a0e233fff31e690cc7a2e837b155b2ee65928229/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/1386a3e884659fbcbb46747234affb859a9d8427/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/167ff4b0ffd2784d4872fde8779653b8/raw/544b8b0e639f2df242696a025d4b38b86afb49ea/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 3]
```diff
@@ -130,13 +137,29 @@ static void uct_ud_mlx5_ep_tx_ctl_skb(uct_ud_ep_t *ud_ep, uct_ud_send_skb_t *skb
...
         dptr            = next_seg;
         wqe_size       += sizeof(*dptr);
         uct_ib_mlx5_set_data_seg(dptr, skb->neth, skb->len, skb->lkey);
     }
 
-    uct_ud_mlx5_post_send(iface, ep, se, ctrl, wqe_size, skb->neth,
+    /* copy IOV from descriptor to WQE */
+    dptr = UCS_PTR_BYTE_OFFSET(ctrl, wqe_size);
+    for (iov_index = 0; iov_index < iovcnt; ++iov_index) {
+        if (!iov[iov_index].length) {
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/ud/accel/ud_mlx5.c]: fixed

File: src/uct/ib/ud/accel/ud_mlx5.h [hunk 4]
```diff
@@ -40,14 +40,17 @@ typedef struct {
 } uct_ud_mlx5_iface_t;
 
 
-static inline unsigned uct_ud_mlx5_tx_moderation(uct_ud_mlx5_iface_t *iface)
+static UCS_F_ALWAYS_INLINE unsigned
+uct_ud_mlx5_tx_moderation(uct_ud_mlx5_iface_t *iface, uint8_t ce_se)
 {
-    if (iface->super.tx.unsignaled >= UCT_UD_TX_MODERATION) {
+    if ((ce_se & MLX5_WQE_CTRL_CQ_UPDATE) ||
+        (iface->super.tx.unsignaled >= UCT_UD_TX_MODERATION - 1)) {
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/ud/accel/ud_mlx5.h]: i guess you've meant to add ( ) around `UCT_UD_TX_MODERATION - 1`  ? added

File: src/uct/ib/ud/base/ud_ep.c (hunk 1 above)
Comment (yosefe) [reply src/uct/ib/ud/base/ud_ep.c]: fixed

File: src/uct/ib/ud/base/ud_iface.c [hunk 5]
```diff
@@ -958,3 +958,9 @@ void uct_ud_iface_progress_enable(uct_iface_h tl_iface, unsigned flags)
 
     uct_base_iface_progress_enable(tl_iface, flags);
 }
+
+void uct_ud_iface_send_completion(uct_ud_iface_t *iface, uint16_t sn,
+                                  int is_async)
+{
+    /* Unimplemented for now */
```
Comment (brminich) [inline src/uct/ib/ud/base/ud_iface.c]: what is the plan?
just checking regarding the am pending ordering

Comment (yosefe) [reply src/uct/ib/ud/base/ud_iface.c]: it does not affect pending queue logic. in addition to increasing send resources in mlx5/verbs, this function would also release some control skb's which could, in turn, call some zero-copy completion callback. all that happens after the message has already been sent.

Comment (brminich) [reply src/uct/ib/ud/base/ud_iface.c]: yes, but can zero-copy completion callback call some tx func?
in this case it would be handled before pending queue

Comment (yosefe) [reply src/uct/ib/ud/base/ud_iface.c]: yes, good point

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 2 of 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 02-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/76932b431d48dd9df4f679a4fd56fa416cd989f1/_main) | :x: FAILURE
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/4c8b01689bfa1549206974fdbc672a67d9a1d74e/hpc-test-node-gpu_W2) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/cb569f9ef85c31bb3bce78d0f958c1b21c2ed82b/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/d69cc389dc884e8f3bf85ae633814a159bc43042/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/4bc9e90fdf3c1e5752c84975e9c5bf9eaf5783dc/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/3c5573bcef52d0cc18a8c3d068ea31de008cddfa/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/fa2df976cd9b19fdbf6c5aab11531b3075400f34/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/972eb933f0560e99fcb2cc4ce03f576bd619c198/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/d8fa4689c4ae8e0492e5fea48194dbccb7c49896/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/d64b90748063a82158ef5eb91119318f523efa76/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/a9999d35320b02354cd02705dae477198c7b1cd1/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/a3bb6210b7bf89141566f9ebd463692351a50192/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/9e88f2effe09bc60dc184ea1f559be5983f23ac1/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/34ed3f17b9909fc50929f22aa4e7a11ab91222bf/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/600b4620f3400d06743070359487ec69e126f69e/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/84bf36257bd3618922447fc47183d3264f0751f2/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/cb759df8845a5b569c5ba32a5f74b60b5885f923/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/271a51566e58fbb9294cfebcd891c40e7ac72a6b/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/6c2b8a5f24a46519bd5d0838206f1fb7c7c3afab/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/eec7654212abb82542eef4a862db03a7c2ece7f6/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/58b0a6bf5e87f1458216d5322e0b38a98d42fb56/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/d0fc178376b92c722fc279d7eb17940c7b00ad82/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/119c524128a02f3c9536d8811b4088c4effebc5c/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/799ea6ea52813b8ee8d35d4caad7474143a6ccdb/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/524f003a3141c85940b9754a28785398/raw/3f4aba4d1a6a7413faed8db0c01508e6d35b1c8a/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by yosefe** [comment]
gpu node failure
bot:mlx:retest

File: src/uct/ib/ud/accel/ud_mlx5.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/uct/ib/ud/accel/ud_mlx5.h]: > i guess you've meant to add ( ) around `UCT_UD_TX_MODERATION - 1` ? added

yep :)

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 02-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/d5cd6a60e84d3e94f848869e82ee4032ed5ad964/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/64d7ec77045d00dd15e2df66c9d06c2927956069/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/edb3512ef1c4392b27e0fb3fc0d339dd9d89086d/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/084f1c769bbeffe0490b10f51119da0b97661f6b/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/192b62acb418e1316f7c9c9154049b90e86242c1/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/8a2a6a5d364bcfa7de0ca5215996c294a211d879/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/fb419a041bf82738fbdfcef5404aaab0906fae6c/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/f9e6428e6a4e96ec001151626f22f108ae284667/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/d997a30d267772a3bae51206abe089445edce773/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/1036d233dc005fdf331a6e57561bd6176a37ec52/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/920cc8b057011b7b67a36cc0f16b7e15e9dc8e8c/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/887f2b76ae3f1cb793faf4b775b69e7bfec30653/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/9fd1ba958fb6cb73156a68dd0dec52ccbef82099/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/0d1b3c2f8fda1a2dd213d3568aa1382c3de5738e/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/8e41e969b23dec3c2e7c1570526d20363f3cec0f/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/8c2e009e550f010b441ec8da8966de6cecd6205f/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/013900e06e97f64f599bba98981d587e5a88de55/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/d4abe92442dbe189481ed7a1e79d8a95c42e34a8/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/4ff871702e25aab116c2bfe8b4ad5d4b645a0f96/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/629c52e7d66b787cbb0f3875d6523a592582e888/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/ba055b707670ee8276577bed8ceb7ff4fa0cad6b/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/1fac1ec5afc3956640c663686797bb5258afe763/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/0c6216db561e6daa8ce7b210c7004bb388132cfc/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/a1ef27376b00be156907bbe94564302a59f5780b/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/8405d57f83da207f38033802aeb6a123/raw/1a851b7e996169c1aebf4f45f16264a7c71a417e/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by yosefe** [comment]
@brminich seems the UD ordering issue is not introduced by this PR (actually, it exists already today: if 2 ACKs are received, the first one increases max_psn, and the send completion callback as a result of 2nd ACK can potentially send data before any queued pending requests)

so can we merge this and fix in next PR?

**Comment by brminich** [comment]
> potentially

You fixed OOO issue with CA window in #4948, but it seems that calling comp functions from new `uct_ud_verbs_iface_poll_tx` func (which is not done yet), would introduce another potential OOO with QP resources

**Comment by yosefe** [comment]
calling comps from poll tx will also be by uct_ud_iface_dispatch_comp, which raises async event, so callbacks are not called from it.

**Pull request merged**